### DISCUSSION TITLE
Support `HelmChartInflationGenerator`  in Kustomization file

### DISCUSSION THREAD

**Discussion by davinkevin** [fluxcd/flux2#2310] [Proposals] [answered]
We would like to use the [`HelmChartInflationGenerator`](https://kubectl.docs.kubernetes.io/references/kustomize/builtins/#_helmchartinflationgenerator_) from `kustomize` with the KustomizeController. 

With this, we are able to do more, especially for `patches`, than using `HelmController` from flux. 
This can also provide single point of entry to deploy everything in multiple env cases (for development, made with `kustomize build . | kubectl apply -f .` and production with `flux`). 

Is there some well known limitation about this or just "adding the helm binary" could do the job?

**Comment by kingdonb** [comment]
If this Kustomize built-in feature shells out to the Helm binary then we cannot use it due to the security posture of Flux.

The Flux design has a policy of no shelling out to binaries. If a function is to be used, then it must be available for use as a go package. So, based on how you've described the `HelmChartInflationGenerator` it will not be possible for Flux to support it.

This posture is important because of the risk profile of compromising Flux – if the `cluster-admin` credential that Flux uses to affect the cluster at the root level is compromised, then it's a total cluster compromise, and potentially worse. So everything about the Flux default security posture assumes we lock down everything, and we should use all available security features for mitigation of this risk.

Flux's first major security audit [was commenced and concluded recently](https://fluxcd.io/security/#advisories), and we were able to mitigate the major CVE issue described there before the audit was over, as adoption of Server Side Apply was already on our roadmap. The vulnerability came from shelling out to `kubectl`, which is no random binary, but just to emphasize – this is where vulnerabilities are most likely to come from, which is the reason for the posture against them.

If kustomize provides a way to inflate helm charts that uses Helm as a go package through well-defined pathways and does not shell out to any random binaries, then we could certainly reconsider!

**Reply by davinkevin** [reply]
Thank you for this complete answer, and I first want to say I a great defender of "no-shell" inside containers 👍.

Currently in official kustomize code, we have that: 
https://github.com/kubernetes-sigs/kustomize/blob/d9435bd1b13a6764b9d271001e61837199494d1c/plugin/builtin/helmchartinflationgenerator/HelmChartInflationGenerator.go#L141-L163

I'll maybe open an issue on Kustomize side, but I don't think they plan to import the complete `helm` library inside `kustomize` (and transitively into `kubectl`)…

One alternative solution could be to provide this "shell out" feature behind a feature flag which is by default set to false. 
Another solution would be to provide an alternative to the `HelmChartInflationGenerator` which would be managed by flux and built in the `kustomizeController`? 

Thank again for your great work 👍

**Comment by ryuheechul** [comment, marked as answer]
My workaround to this issue is as following:

let's say it's at `./kustomize`

```bash
cd kustomize

mkdir with-inflator
mv *.yaml with-inflator

mkdir hard-coded
cd hard-coded
kustomize create

# you should need to run this every time you upgrade the chart
kustomize build --enable-helm ../with-inflator > chart-rendered.yaml

cat <<EOF >> kustomization.yaml
resources:
- ./chart-rendered.yaml
EOF
```

and now update the `path:` in your `clusters/[your-cluster]/blahblah-kustomization.yaml` to `./kustomize/hard-coded`

(and commit and push)

**Reply by kingdonb** [reply]
Yes, and – if you do this in your CI system, it can write the result back to git, so it can be both inspected after templates are rendered, and applied directly by Flux's kustomize controller in the usual way.

This is great, I think this is the real answer. 🏅 the only drawback for me is, the rendered chart in a single file. I'd like to splat that file back out into folders that are named something more like `namespace/resource-kind.yaml`. I may write a quick Ruby script or Go utility (Ruby is my preference...) to add to this, so that the chart output can be easier to inspect and check for diffs.

**Reply by kingdonb** [reply]
Here we go, I've added this module to a Ruby project I've been working on that generates manifests from a Ruby parsing of YAML that comes through another Ruby process... yay Ruby (ref: https://github.com/kingdonb/scrob-web/pull/8)

https://github.com/kingdonb/scrob-web/pull/8/files#diff-bb444aacbe95e047c39244be84186b222df5b7c98cac55c157f9a84b00fde66fR47-R53

Walkthrough: the `Splat` module has two functions it provides (the rest should not be public, but Ruby does not make it very easy to write private module functions iirc so whatever)

```ruby
Splat.write_namespaces_into_folder(target_dir: './secrets',
    namespaces_arr: Splat.sort_into_namespaces(secrets))
```

The functions are `write_namespaces_into_folder` and `sort_into_namespaces` and I wrote [tests](https://github.com/kingdonb/scrob-web/blob/26ea43c2a1015dabf01a74e8045373a8fc7b1beb/spec/splat/splat_spec.rb) to show what they can do (and to roughly cover the limits of things which I have expected that they can or should not do.) If you can use Ruby, maybe we extract this into a script that runs as a single file? I had already this pipeline of Ruby stuff going, so there was a natural place for me to plug in this new module, and I went ahead and did it as a test.

The results are explained [in the body of the PR](https://github.com/kingdonb/scrob-web/pull/8). You can take a stream of YAML (or a multi-doc YAML file) as input, and generate as output a directory full of `ns-dir/object-name-kind.yaml` which seems like a really easy problem, but has a few gotchas, including that Kustomize sometimes applies namespaces for you later on down the line, so while Flux requires namespaces, your YAML docs might better omit them and let a Kustomize Overlay apply the namespaces later on... so the script had better accommodate both cases.

Anyway, I wrote some tests and if you're keen, you can [translate this into Go](https://github.com/kingdonb/scrob-web/blob/26ea43c2a1015dabf01a74e8045373a8fc7b1beb/lib/splat/splat.rb), or I probably will in a little while! I just knew that I could write it much faster in Ruby (and test it more completely and effectively on my own in Ruby than in Go) and I was already using Ruby for the YAML parser, so here we are. 🚀 🎸 🤠 👍 

The script linked above demonstrates how to splat a single file or stream input into a directory of directories of files like Flux expects!

Edit: https://github.com/kingdonb/scrob-web/blob/main/input-splitter.rb - here's a simplified version of the script, with less entanglement in my other stuff – doesn't hook into Kuby or any of this other stuff. It just calls the library mentioned above and reads a stream of YAML docs from STDIN then turns them into a directory full of files.
